### PR TITLE
Sol-list: Add test suite

### PR DESCRIPTION
**Pull request by kaspersky**
#1661

### CODE REVIEW

**Comment by kaspersky** [comment]
The SOL_LIST_INIT segfault comment was too vague, so since sol-list resembles linux kernel lists implementation very much I made it similar (there is now a usage of the macro in the list test).

**Comment by cabelitos** [comment]
lgtm

**Comment by barbieri** [comment]
@cabelitos I guess he doesn't have commit rights, so would you rebase/push?

File: src/lib/datatypes/include/sol-list.h [hunk 1]
```diff
@@ -66,9 +66,15 @@ struct sol_list {
...
+ * @brief Helper macro to initialize a @c sol_list structure. Example:
+ *
+ * @code
+ * struct sol_list mylist = SOL_LIST_INIT(mylist);
+ * @endcode
+ *
+ * @param name List to be initialized
  */
-#define SOL_LIST_INIT { NULL, NULL }
+#define SOL_LIST_INIT(name) { &(name), &(name) }
```
Comment (ceolin) [inline src/lib/datatypes/include/sol-list.h]: This is partially right. This is a kind of inlist, it means that this struct should be used in other struct to provide lists. There is no utiliity in use  it directly, as you can see the struct has two fields `prev` and `next` that with this macro will point to the same address. The usage of this is:

```
struct delayed_packet {
    struct sol_list list;
    struct sol_flow_packet *packet;
    uint16_t source_idx;
    uint16_t source_port_idx;
};
```

It said, i guess it is better initialize with the value given by the user and not get the address.

File: src/test/test-list.c [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        ASSERT_INT_EQ(entry->data, i++);
+
+        sol_list_remove(it);
+        free(entry);
+    }
+
+    ASSERT(sol_list_is_empty(&list2));
+}
+
+TEST_MAIN();
```
Comment (ceolin) [inline src/test/test-list.c]: :+1:  for test. Nice job

File: src/lib/datatypes/include/sol-list.h (hunk 1 above)
Comment (kaspersky) [inline src/lib/datatypes/include/sol-list.h]: @ceolin, I don't fully understand what you mean. As you can see in the test, the usage is the same as the code you provided. Following your example, you need a sentinel (list variable) to access the list API:

```c
struct sol_list delayed_packets = SOL_LIST_INIT(delayed_packets);

sol_list_add(&delayed_packets, &delayed_packate.list);
```

The sentinel must have ```next``` and ```prev``` pointers set to its address (that is the convention for an empty list). The limitation of ```SOL_LIST_INIT``` is that you have to pass a sol_list struct, and not a pointer to it, in the last case you have ```sol_list_init```. If you mean to define the macro as ```{ (name), (name) }```, then you'll have to have something like ```*my_list = SOL_LIST_INIT(my_list);``` which looks strange.

Comment (cabelitos) [inline src/lib/datatypes/include/sol-list.h]: @ceolin I did not understand as well. @kaspersky approach looks fine to me

Comment (ceolin) [inline src/lib/datatypes/include/sol-list.h]: ok, go ahead with this

**Comment by ceolin** [comment]
merged :)

**Pull request closed without merging**